File: 78-Subsets/78-Subsets.java
// Last updated: 8/16/2026, 9:52:49 PM
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        // DFS
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        helper(nums, res, new ArrayList<>(), 0);
        return res;
    }

    private void helper(
        int[] nums, 
        List<List<Integer>> res, 
        List<Integer> subset, 
        int idx
    ) {
        // find all subsets starting with 'subset' 
        // next available number is nums[idx]

        res.add(new ArrayList<Integer>(subset)); //deep copy subset

        for(int i = idx; i < nums.length; i ++) {
            subset.add(nums[i]);
            helper(nums, res, subset, i + 1);
            subset.remove(subset.size() - 1);
        }

    }
}

File: 78-Subsets/78-Subsets.py
# Last updated: 8/23/2026, 11:32:16 AM
1class Solution:
2    def subsets(self, nums: List[int]) -> List[List[int]]:
3        # backtrack
4        res = []
5        path = []   # current path
6
7        # find subsets in nums[start:]
8        def backtrack(start: int):
9            # every path is legit
10            res.append(path[:])     # copy of path (cause path is changing)
11
12            for i in range(start, len(nums)):
13                path.append(nums[i])
14                backtrack(i + 1)    # num cannot be reused, so next is i+1
15                # cancel
16                path.pop()
17
18
19        backtrack(0)
20        return res